bioconda_utils/build.py

```diff
@@ -0,0 +1,351 @@
+                docker_builder=docker_builder
+            )
+
+            all_success &= target_success
+            recipe_success &= target_success
+
+            if not target_success:
+                failed.append((recipe, target))
+                for n in nx.algorithms.descendants(subdag, name):
+                    skip_dependent[n].append(recipe)
+        if recipe_success:
+            built_recipes.append(recipe)
+
+    if len(failed) > 0:
+        failed_recipes = list(set(i[0] for i in failed))
+        logger.error(
+            'BIOCONDA BUILD SUMMARY: of %s recipes, '
+            '%s failed and %s were skipped. '
+            'Details of recipes and environments follow.',
+            len(recipes), len(failed_recipes), len(skipped_recipes))
+
+        logger.error(
+            'BIOCONDA BUILD SUMMARY: while the entire build failed, '
+            'the following recipes were built successfully:\n%s',
+            '\n'.join(built_recipes))
+
+        for recipe, target in failed:
+            logger.error(
+                'BIOCONDA BUILD SUMMARY: FAILED recipe %s, environment %s',
+                str(target), target.envstring())
+
+        for name, dep in skip_dependent.items():
+            logger.error(
+                'BIOCONDA BUILD SUMMARY: SKIPPED recipe %s '
+                'due to failed dependencies %s', name, dep)
+        return False
+
+    logger.info(
+        "BIOCONDA BUILD SUMMARY: successfully built %s of %s recipes",
+        len(built_recipes), len(recipes))
+
+    if not testonly:
+        # upload builds
+        if (os.environ.get("TRAVIS_BRANCH") == "master" and
+                os.environ.get("TRAVIS_PULL_REQUEST") == "false"):
+            for recipe in recipes:
+                for target in recipe_targets[recipe]:
+                    package = target.pkg
+                    logger.debug(
+                        "Checking existence of package {}".format(package)
+                    )
+                    if os.path.exists(package):
+                        logger.info("Uploading package {}".format(package))
+                        try:
+                            sp.run(
+                                ["anaconda", "-t",
+                                 os.environ.get("ANACONDA_TOKEN"), "upload",
+                                 package],
+                                stdout=sp.PIPE,
+                                stderr=sp.STDOUT,
+                                check=True)
+                        except sp.CalledProcessError as e:
+                            print(e.stdout.decode(), file=sys.stderr)
+                            if b"already exists" in e.stdout:
+                                # ignore error assuming that it is caused by
+                                # existing package
+                                pass
+                            else:
+                                raise e
+                    else:
+                        logger.error(
```

Comment:
Will do.

Also note that this code path is totally untested at the moment. I'd like to set up a bioconda-utils channel with full read/write anaconda client access. Once the encrypted ANACONDA_TOKEN is available in the .travis.yaml, we can test package uploading (and then clean up when done). Without such a test channel, we would have to wait for merge into bioconda-recipes master to learn that the uploading fails.